lib/Dialect/Pipeline/PipelineOps.cpp

```diff
@@ -15,13 +15,53 @@
 #include "mlir/Dialect/Func/IR/FuncOps.h"
 #include "mlir/IR/Builders.h"
 #include "mlir/IR/FunctionImplementation.h"
+#include "llvm/Support/Debug.h"
 
 using namespace mlir;
 using namespace circt;
 using namespace circt::pipeline;
 
 #include "circt/Dialect/Pipeline/PipelineDialect.cpp.inc"
 
+#define DEBUG_TYPE "pipeline-ops"
+
+// NOLINTNEXTLINE(misc-no-recursion)
```

Comment:
This one seems easy to convert to non-recursive. Plz do so and remove this line. LLVM and/or MLIR doesn't by default allow recursion for a reason.